File: go/meetup/meetup.go
package meetup

import (
	"time"
)

const testVersion = 3

// WeekSchedule represents how to schedule a meetup
type WeekSchedule int

// WeekSchedule values
const (
	First WeekSchedule = iota
	Second
	Third
	Fourth
	Last
	Teenth
)

// Day returns the date of the scheduled meetup
func Day(s WeekSchedule, w time.Weekday, m time.Month, y int) (day int) {
	switch s {
	case Teenth:
		date := time.Date(y, m, 13, 0, 0, 0, 0, time.UTC)
		day = date.AddDate(0, 0, daysUntilWeekday(date, w)).Day()
	case Last:
		date := time.Date(y, m+1, 1, 0, 0, 0, 0, time.UTC)
		date = date.AddDate(0, 0, -1)
		day = date.AddDate(0, 0, -daysAfterWeekday(date, w)).Day()
	default:
		date := time.Date(y, m, 1, 0, 0, 0, 0, time.UTC)
		day = date.AddDate(0, 0, daysUntilWeekday(date, w)+int(s)*7).Day()
	}

	return
}

func daysUntilWeekday(date time.Time, w time.Weekday) int {
	return (7 + int(w-date.Weekday())) % 7
}

func daysAfterWeekday(date time.Time, w time.Weekday) int {
	return (7 + int(date.Weekday()-w)) % 7
}

File: ruby/meetup/meetup.rb
# frozen_string_literal: true

require 'date'

class Meetup
  OCCURENCES = %i[first second third fourth].freeze
  WEEKDAYS = %i[sunday monday tuesday wednesday thursday friday saturday].freeze

  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, occurence)
    case occurence
    when :teenth
      date = Date.new(@year, @month, 13)
      date + days_until_weekday(date, weekday)
    when :last
      date = Date.new(@year + @month / 12, @month % 12 + 1, 1) - 7
      date + days_until_weekday(date, weekday)
    else
      date = Date.new(@year, @month, 1)
      date + days_until_weekday(date, weekday) + OCCURENCES.index(occurence) * 7
    end
  end

  def days_until_weekday(date, weekday)
    (7 + WEEKDAYS.index(weekday) - date.cwday) % 7
  end
end

module BookKeeping
  VERSION = 1
end
